### ISSUE TITLE
transform the return value of `Result` and `Option` type

### ISSUE THREAD

**Issue by nerditation** [Kobzol/rust-delegate#59]
I've read #48 where attributes like `#[method(clone)]` was discussed.

I have a specific use case, where I have a wrapper type and want to delegate a lot of method to the inner field. but I'd like to convert the return Result to my error type, something like:

```rust
struct Wrapper(Foo);
impl Wrapper {
    // map_err() can be used
    pub fn bar(&self) -> Result<i32, MyError> {
        self.0.bar().map_err(From::from)
    }
    // can also use the question mark operator
    pub fn bar(&self) -> Result<i32, MyError> {
        Ok(self.0.bar()?)
    }
    // sometimes the inner type return Option, convert to Result
    pub fn bar(&self) -> Result<i32, MyError> {
        self.0.bar().ok_or(MyError::EmptyResponse)
    }
    // I didn't have this, but for completeness, convert Result to Option
    pub fn bar(&self) -> Option<i32> {
        self.0.bar().ok()
    }
}
```

it would be nice if `delegate` can support this. especially for the question mark case

**Comment by Kobzol** [comment]
Question mark is common enough I suppose to warrant a special attribute. But at this point, I wonder if we should just create some generic expr wrapper, like
```rust
struct Wrapper(Foo);
impl Wrapper {
    delegete! {
        #[expr(Ok($expr?))]
        to self.0 {
             fn foo(&self) -> Result<..., ...>;;
             fn bar(&self) -> Result<..., ...>;
        }
    }
}
```

**Comment by Kobzol** [comment]
I guess that it will depend on what will be the easiest to parse with `syn`, I'm not even sure if we can start the sigil with `$`. I'm leaving for a two week vacation now, so I won't be able to work on this now, I'll experiment with it in September.

**Comment by Kobzol** [comment]
Hmm, `$` nor `$expr` can be parsed as an expression with `syn`. And also I realized that this would clash with the `to <expr>` target expression a little bit. Ideally we would be able to say something like `delegate! { to Ok(self.foo.$call) }`, but this conflates two things - the target of the call (`self.foo`) and an additional "wrapping" expression around the resulting call (`Ok(...)`). And also this syntax wouldn't work, because sometimes the call can be performed with `::` instead of `.` (for associated functions).

**Comment by nerditation** [comment]
> delegate! { to Ok(self.foo.$call) }

yeah, too much is going into a single syntax form, and that looks complicated to me. it would be good for a general solution, but I think for now, make the question mark operator a special case is acceptable.

my first thought was maybe we can decorate the function signature a little bit, similar to how function parameters can be transformed in current version, something along the lines of:

```rust
delegate! {
    to self.0 {
        // this syntax?
        pub fn foo(&self, #[into] other: Self) -> #[Ok(?)] Result<isize, Self::Error>;
        // or this?
        pub fn foo(&self, #[into] other: Self) -> #[map_err] Result<isize, Self::Error>;
        // or even this?
        pub fn foo(&self, #[into] other: Self) -> ? Result<isize, Self::Error>;
        // but it feels too complex if we want to fit too much features
        pub fn foo(&self, #[into] other: Self) -> #[ok_or(Error::DataEmpty)] Result<isize, Self::Error>;
    }
}
```

looking into the documents, I find that currently we support `#[unwrap]`, `#[try_into]`, `#[into]`, I think it would be reasonable to add the opposite of `#[unwrap]`, I don't have a suitable naming yet, maybe something like `#[Ok]` and `#[Some]`, (and by extension, an variant form like `#[Ok(?)]`, although it would be useless for a `#[Some(?)]`), this should coverage the common use cases of the form `Ok(self.inner.foo()?)` I think.

**Comment by Kobzol** [comment]
I'm still unsure about this, as I don't like piling new and new attributes into this crate. I think that some experimentation with the placeholder syntax (like `#[wrap(Ok(_)?)]`) could be nice, but I don't currently have time to play with it. I'd be happy to accept PRs where we could discuss a specific implementation though :)

**Comment by Kobzol** [comment]
Ok :) Closing for now.

**Comment by vic1707** [comment]
Continuing #80.

@Kobzol are you open to adding 2 attributes for this or 1 is the maximum?
I kinda like the idea of doing a `#[wrap(...)]` and possibly a `#[then(...)]` placed above function declaration.

I think this is doable 
#### `#[then(...)]` as
- `#[then(toto)]` => `$expr.toto()`
- `#[then(toto(<args>))]` => `$expr.toto(<args>)`

#### `#[wrap(...)]` as
- `#[wrap(toto)]` => `toto($expr)`
- `#[wrap(toto(<args>))]` => `toto($expr, <args>)`
---

Also, I don't think supporting `?` is needed (at least not right away) as
```rs
pub fn bar(&self) -> Result<i32, MyError> {
    Ok(self.0.bar()?)
}
```
can simply be written 
```rs
pub fn bar(&self) -> Result<i32, MyError> {
    self.0.bar()
}
```
or function throwing multiple times
```rs
pub fn bar(&self) -> Result<i32, MyError> {
    self.0.bar()?.baz()?.toto()
}
```
would need a `#[then(...)]` attribute, or I'm missing a use-case 🤔

**Comment by Kobzol** [comment]
I'm still not convinced that this is worth it. This crate already provides several special cases for various common things, but having a completely custom syntax would mean that we would essentially be reinventing just writing Rust. If the expression is so complicated, I would just write out the method explicitly.

**Comment by vic1707** [comment]
> This crate already provides several special cases for various common things

Having a unified way of wrapping & calling methods for each arm makes it possible to remove (or turn into aliases) `#[unwrap]`/`#[into]`/`#[try_into]`/`#[as_ref]` and possibly `#[new_type]` if `#[then(...)]` takes in an expression to append eg: `#[then(.0)]` &  `#[then(.unwrap())]`.
Wouldn't that simplify the `delegate` api ?

> but having a completely custom syntax would mean

I’m not sure I understand this concern. Are you referring to:
```rs
#[then(...)]
#[wrap(...)]
#[const(...)]
pub fn toto(&self) -> ..;
```
Or is there another issue I’m missing?

P.S. I hope this doesn’t come across as trying to push anything—just making sure I fully understand your position and concerns. 🙇

**Comment by Kobzol** [comment]
> As for ? I really think it's unnecessary (see my point above) but I get why newcomers could ask for it.

It's not unnecessary if you need the `From` error conversion that ? does :)

> #[expr(to_toto($expr?.unwrap().into()))]

Yes, this is something that I had in mind. It's quite complex, but at least we would have a single approach of how to specify such complex expression. It's much better than
```
#[wrap(to_toto)]
#[then(?)]
#[then(unwrap)]
#[then(into)]
```

**Comment by Kobzol** [comment]
This can be now achieved using the generic `#[expr]` attribute (https://github.com/Kobzol/rust-delegate/pull/82).
